updates_and_signals/safe_message_handlers/workflow.py

```diff
@@ -228,29 +193,16 @@ def should_continue_as_new(self) -> bool:
     @workflow.run
     async def run(self, input: ClusterManagerInput) -> ClusterManagerResult:
         self.init(input)
-        await workflow.wait_condition(lambda: self.state.cluster_started)
-        # Perform health checks at intervals.
-        while True:
-            await self.perform_health_checks()
-            try:
-                await workflow.wait_condition(
-                    lambda: self.state.cluster_shutdown
-                    or self.should_continue_as_new(),
-                    timeout=timedelta(seconds=self.sleep_interval_seconds),
-                )
-            except asyncio.TimeoutError:
-                pass
-            if self.state.cluster_shutdown:
-                break
-            if self.should_continue_as_new():
-                workflow.logger.info("Continuing as new")
-                workflow.continue_as_new(
-                    ClusterManagerInput(
-                        state=self.state,
-                        test_continue_as_new=input.test_continue_as_new,
-                    )
-                )
-        return ClusterManagerResult(
-            len(self.get_assigned_nodes()),
-            len(self.get_bad_nodes()),
+        await workflow.wait_condition(
+            lambda: self.state.cluster_shutdown or self.should_continue_as_new()
         )
+        if self.should_continue_as_new():
```

Comment:
Right, I'm trying to move too quickly here. Reinstated loop, and added wait for `all_handlers_finished()` before CAN. It would be ideal to add some tests exercising CAN.